background/style-manager.js

```diff
@@ -526,19 +533,40 @@ const styleManager = (() => {
     return '^' + escapeRegExp(text).replace(/\\\\\\\*|\\\*/g, m => m.length > 2 ? m : '.*') + '$';
   }
 
-  function getDomain(url) {
-    return url.match(/^[\w-]+:\/+(?:[\w:-]+@)?([^:/#]+)/)[1];
-  }
-
-  function getUrlNoHash(url) {
-    return url.split('#')[0];
-  }
-
   // The md5Url provided by USO includes a duplicate "update" subdomain (see #523),
   // This fixes any already installed styles containing this error
   function fixUsoMd5Issue(style) {
     if (style && style.md5Url && style.md5Url.includes('update.update.userstyles')) {
       style.md5Url = style.md5Url.replace('update.update.userstyles', 'update.userstyles');
     }
   }
+
+  function createMatchQuery(url) {
+    let urlWithoutHash;
+    let urlWithoutParams;
+    let domain;
+    return {
+      url,
+      get urlWithoutHash() {
+        if (!urlWithoutHash) {
+          urlWithoutHash = url.split('#')[0];
+        }
+        return urlWithoutHash;
+      },
+      get urlWithoutParams() {
+        if (!urlWithoutParams) {
+          const u = new URL(url);
```

Comment:
I'm not sure if this is a safe method when using non-http URLs e.g. `about:blank`, `file:///`, etc. It may work differently in different browsers.